packages/android/barnard/src/main/kotlin/org/levarac/barnard/BarnardParticipantRelay.kt

```
@@ -0,0 +1,111 @@
+) {
+    private data class Candidate(val digest: ByteArray, val envelope: ByteArray, var hop: Int, val eventId: ByteArray, val verifiedAt: Long, val validFrom: Long, val validThrough: Long, val expires: Long)
+    private val electionKey = randomnessSeedMaterial.copyOf()
+    private val candidates = linkedMapOf<String, Candidate>()
+    private var selected: String? = null; private var pinUntil = 0L
+    private val handles = linkedMapOf<String, Long>()
+    private var overflowSeenAt: Long? = null
+    private var activeUntil: Long? = null; private var contentionUntil: Long? = null; private var lastDecisionEpoch: Long? = null; private var stopped = false
+
+    private fun withinRelayWindow(c: Candidate, current: Long): Boolean = current in c.validFrom until c.expires
+
+    public fun observe(container: ByteArray, peerHandle: ByteArray): BarnardRelayObservationResult {
```

detekt (1.23.8)

[warning] The function observe appears to be too complex based on Cyclomatic Complexity (complexity: 25). Defined complexity threshold for methods is set to '25'

(detekt.complexity.CyclomaticComplexMethod)